### PR TITLE
Improved Reference Metric and SIMD

### PR DESCRIPTION
**Pull request by ksible**
I simplified the algorithm for reference_metric.py and added backward-compatibility since SymPy changed the derivative representation with SymPy 1.2. I implemented an expression tree data structure (SIMDExprTree.py) for direct manipulation of the SymPy expression tree with better performance than the build-in xreplace function. I improved the performance of the SIMD algorithm (SIMD.py) for NRPy using the previously mentioned expression tree.

### CODE REVIEW

File: SIMD.py [hunk 1]
```diff
@@ -222,89 +239,80 @@ def IntegerPowSIMD(a, n):
...
-        if item.func == AddSIMD and item.args[1].func == MulSIMD:
-            tempitem = FusedMulAddSIMD(item.args[1].args[0],item.args[1].args[1],item.args[0])
-
-        # If the pattern is a*b - c, replace with FMS(a,b,c)
-        if item.func == SubSIMD and item.args[0].func == MulSIMD:
-            tempitem = FusedMulSubSIMD(item.args[0].args[0],item.args[0].args[1],item.args[1])
-
-        if item != tempitem: expr = expr.subs(item, tempitem)
-
-    # Step 7: SIMD intrinsics cannot take integers or rational numbers as arguments.
```
Comment (zachetienne) [inline SIMD.py]: I think you chopped off this comment. Please add it back, revise if need be.

File: reference_metric.py [hunk 2]
```diff
@@ -626,38 +626,23 @@ def ref_metric__hatted_quantities(SymPySimplifyExpressions=True):
...
-                        deriv_op += "D"
-                    deriv_op += deriv_wrt
-                    for i in range(deriv_order - 1):
-                        deriv_op += deriv_wrt
-                    rule[item] = sp.sympify(strfunc + deriv_op)
+                    if outdated_version:
+                        var, order = str(item.args[1])[2:], len(item.args) - 1
+                    else:
+                        var, order = str(item.args[1][0])[2:], item.args[1][1]
+                    oper = '__D' + 'D'*(order - 1) + var*order
```
Comment (zachetienne) [inline reference_metric.py]: Please describe what is being done here. Provide an example input & output.

File: SIMDExprTree.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
+__author__ = 'Ken Sible'
```
Comment (zachetienne) [inline SIMDExprTree.py]: Please add a header consistent with other NRPy+ Python modules that contains author info and briefly describes the purpose of the module.

**Comment by ksible** [comment]
I fixed the issue with Travis-CI failing on Python 2. The error that kept arising was "only named arguments may follow *expression," which occurs whenever a non-keyword argument follows tuple unpacking. I removed the tuple unpacking from that failed section of the SIMD script. Furthermore, the string interpolation in SIMDExprTree was replaced with string formatting for compatibility.

**Pull request merged**